standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/DecimalComparator.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.metastore.columnstats;
+
+import org.apache.hadoop.hive.metastore.api.Decimal;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Comparator;
+import java.util.function.Consumer;
+
+public class DecimalComparator implements Comparator<Decimal> {
+
+  public static final DecimalComparator INSTANCE = new DecimalComparator();
+
+  public static final byte PAD_POS = (byte) 0;
+  public static final byte PAD_NEG = (byte) -1; // or 0xff in hexadecimal
+
+  public enum Approach {
+    UNKNOWN, SIGN, EQSCALE, LOG_ZERO, BITLOG_A, BITLOG_B, FALLBACK;
+
+    public static final int LEN = Approach.values().length;
+  }
+
+  /**
+   * Notifies the caller which approach was chosen.
+   *
+   * <p>It is used by tests.
+   */
+  interface ApproachInfoCallback extends Consumer<Approach> {
+  }
+
+  @Override
+  public int compare(Decimal o1, Decimal o2) {
+    return compareInner(o1, o2, true, null);
+  }
+
+  private boolean isNotNegative(byte[] unscaled) {
+    if (unscaled.length == 0) {
+      return true;
+    }
+    return 0 == (Byte.toUnsignedInt(unscaled[0]) >> 7);
+  }
+
+  int compareInner(Decimal d1, Decimal d2, boolean useFallback, ApproachInfoCallback aic) {
+    byte[] unscaled1 = d1.getUnscaled();
+    byte[] unscaled2 = d2.getUnscaled();
+    boolean notNegative1 = isNotNegative(unscaled1);
```

Comment:
about naming on overall: isn't it better to understand `isPositive` instead of `isNotNegative`?